Fix 4472 - Display notice when DelayJS is activated with Autoptimize Aggregate JS already in effect

iCaspar opened the pull request:
## Description

Please include a summary of the change and which issue is fixed/closed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes #4472
Fixes #3707 

## Type of change
- [X] Enhancement (non-breaking change which improves an existing functionality)

## Is the solution different from the one proposed during the grooming?

Yes
In the grooming I had thought we would be adding a helper notice into the DelayJS selection field. In fact, we are adding a notice to the notices area at the top of the page.
Also, to do this, we had been adding the notice via the '`settings_errors` transient,  but that was failing to display messages properly. Instead, this adds a method in the DelayJS Admin/Subscriber to check if Autoptimize Aggregate JS is active, and if so, uses `add_settings_errors()` for display.

## How Has This Been Tested?

- [X] New Integration tests to confirm that the notice is added under expected conditions.
- [x] Check that notice is displayed properly on local environment.

<img width="1107" alt="Screen Shot 2021-11-17 at 9 32 55 AM" src="https://user-images.githubusercontent.com/7462998/142220779-d69a3e30-87cc-4078-aa7c-755967d6a9ad.png">

# Checklist:

Please delete the options that are not relevant.

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

iCaspar committed:
Add tests and method for AO agg-js and wpr delayjs

iCaspar committed:
Fix param order in docblock

iCaspar commented:
This approach should work, as the integration tests prove that notice data is added to the `settings_errors` transient array. But the notice is not appearing, due to the same bug that is causing https://github.com/wp-media/wp-rocket/issues/4314, namely there is something preventing any notice being displayed from the `settings_errors` transient when settings are updated.

Marking this as blocked by #4314.

iCaspar committed:
Use add_settings_error instead of transient

iCaspar committed:
Fix docblock order

iCaspar committed:
Remove dead transient_errrors code

iCaspar committed:
Remove extra spacing

iCaspar committed:
Use matching priority in test

iCaspar committed:
Remove using void thing.

Mai-Saad commented:
@iCaspar Please find notes below based on test analysis and exploratory test:

1- If Delay JS was enabled, then we enable auto-optimize with the aggregate JS features => no warning will be displayed (we will need to re-enable delay js to see the warning), can we display the warning without needing to re-enable Delay JS (maybe with cache clear, refresh settings)?

2- When Autooptimize aggregate js files and inline js is enabled along with WPR LL video and preview image, result in removing LL videos script from HTML => Do we need a separate GH issue for that? or is it something we can handle on this PR? (same already on trunk)

3- @piotrbak  Q: Why not handle the current conflict, the same way as we do with minify/combine CSS/JS (disable the option and display red msg)?
![Screenshot from 2021-11-18 12-27-13](https://user-images.githubusercontent.com/76941962/142400280-b8c2ff6b-af5d-4621-b499-ecfa09f528b8.png)

4- Just for UI confirmation => Current msg which is displayed only when enable delay js while aggregate options are enabled in auto-optimize is as attached, is this what we need?
![Screenshot from 2021-11-18 12-28-03](https://user-images.githubusercontent.com/76941962/142400760-2f3ab177-52ce-4788-b093-a506c8b0f409.png)

piotrbak commented:
@Mai-Saad **When it comes to the 3:**
Minify and Combine CSS/JS are exactly the same features. It's not the case with DelayJS, as it's not causing errors, rather not working as good as it should, we decided to go with a educative warning approach.

@iCaspar Can we make the warning border yellow instead of blue?

piotrbak commented:
@Mai-Saad 

**For 2:**
Does it cause any issues with the video or lazyload?

**For 4:**
We're going to display the message when:
1. Delay JS is enabled
2. Autoptimize JS Aggregation is enabled

CSS files are disabled because the CSS optimisations are enabled in the Autoptimize, below we have an explanation

iCaspar commented:
@piotrbak @Mai-Saad 
On 1, please confirm that the User Story above is accurate. Per the current version of the story:
We will show the message when Autoptimize>JS Aggregation is already active and then DelayJS is activated (the behavior Mai observed). If we also want it to appear when the events happen in the reverse order (DelayJS is already active and then JSAggregation is activated) we should specify that behavior in the story. Otherwise, I think it is working as intended here.

On 2, interaction with Lazyload -- this PR is ONLY about messaging -- see the User Story. It will not include any fixes for conflicts with LazyLoad (or anything else).

On 3, see 2.

On 4, the message is (again) specified in the User Story.

piotrbak commented:
@iCaspar Will it be possible to cover both scenarios? We want to display the warning in all possible use cases. The same should be applied to other Autoptimize compatibilities. Sorry for the misunderstanding.

iCaspar commented:
@piotrbak No problem. Let's update the User Story and I can make that change this morning.

iCaspar committed:
Move warning method to new Compat Subscriber

iCaspar committed:
Reconfigure for new story conditions

iCaspar committed:
Add Autoptimize class to container

iCaspar committed:
Remove dead code from DelayJs Subscriber

iCaspar committed:
Move test to ThirdParty

iCaspar committed:
Check failing test

This will display the notice as expected, but the integration test will not run in isolation. Other functions from preload still run on admin_notice.

iCaspar committed:
Fix phpcs

iCaspar committed:
Revise tests and make pass

engahmeds3ed commented on inc/ThirdParty/Plugins/Optimization/Autoptimize.php:
```diff
@@ -0,0 +1,106 @@
+	 * @since 3.10.4
+	 */
+	public function warn_when_js_aggregation_and_delay_js_active() {
+		if ( ! rocket_get_constant( 'AUTOPTIMIZE_PLUGIN_VERSION', false ) ) {
+			return;
+		}
```
I think it would be better adding this check into `get_subscribed_events` method not to register the file events if the plugin is not installed, this will help if we added other events in the future.

iCaspar replied:
Initially, I did have it where you are suggesting. However, when `get_subscribed_events()` returns and empty array, that was what seemed to cause the failure in the `removeAllCallbacksExcept()` from detatching the other things from `admin_notices`. I think we will have to leave it here for now, until we can fix the callbacks test method.

engahmeds3ed commented on inc/ThirdParty/Plugins/Optimization/Autoptimize.php:
```diff
@@ -0,0 +1,106 @@
+	 * @since 3.10.4
+	 */
+	public function warn_when_js_aggregation_and_delay_js_active() {
+		if ( ! rocket_get_constant( 'AUTOPTIMIZE_PLUGIN_VERSION', false ) ) {
```
`AUTOPTIMIZE_PLUGIN_VERSION` is introduced at autoptimize version 2.4.0 (3 years ago) and I think it wasn't there before, is this a problem or we are fine? @wp-media/productrocket

piotrbak replied:
I think we're fine, I don't believe anyone will be using 3 years old Autoptimize + newest WP Rocket

iCaspar replied:
This is OK, per @piotrbak in slack.

engahmeds3ed reviewed (changes requested):
Small minor comments

iCaspar committed:
Fix open/close strong tags

iCaspar committed:
Fix strong tags in test

iCaspar committed:
Remove extra dismiss icon

Mai-Saad reviewed (commented):
Working as expected.

As per discussion, we agreed on the following notes:
- Notice won't be displayed after dismissing with plugin(s) reactivation (with same settings), import same settings
- Notice will be normally displayed when delay js is enabled independent of exclusions (i.e with default exclusions, notice is displayed)
- When enabling autoptimize while LL video is enabled, LL video will be automatically unchecked (this is already on the trunk and added ticket for this)
[testrail-report-311.pdf](https://github.com/wp-media/wp-rocket/files/7589642/testrail-report-311.pdf)
